File: PoTicTac.Client/Services/GameLogicService.cs
using PoTicTac.Client.Models;

namespace PoTicTac.Client.Services;

public class GameLogicService
{
    public GameBoardState CreateInitialState(PlayerType startingPlayer, Player[] players)
    {
        var initialState = new GameBoardState
        {
            CurrentPlayer = startingPlayer,
            GameStatus = GameStatus.Playing,
            Winner = null,
            Players = players,
            MoveHistory = new List<Move>(),
            UndoStack = new List<Move>(),
            RedoStack = new List<Move>(),
            SessionStats = new SessionStats
            {
                TotalGames = 0,
                StartTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                LastUpdateTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
            }
        };

        // The constructor already initializes the board to all None values
        return initialState;
    }

    public PlayerType? CheckWinner(GameBoardState board)
    {
        const int winLength = 4; // Need 4 in a row to win
        int boardSize = GameBoardState.BoardSize;

        // Check rows for 4 in a row
        for (int row = 0; row < boardSize; row++)
        {
            for (int col = 0; col <= boardSize - winLength; col++)
            {
                PlayerType first = board.Board[row, col];
                if (first != PlayerType.None)
                {
                    bool win = true;
                    for (int i = 1; i < winLength; i++)
                    {
                        if (board.Board[row, col + i] != first)
                        {
                            win = false;
                            break;
                        }
                    }
                    if (win)
                    {
                        return first;
                    }
                }
            }
        }

        // Check columns for 4 in a row
        for (int col = 0; col < boardSize; col++)
        {
            for (int row = 0; row <= boardSize - winLength; row++)
            {
                PlayerType first = board.Board[row, col];
                if (first != PlayerType.None)
                {
                    bool win = true;
                    for (int i = 1; i < winLength; i++)
                    {
                        if (board.Board[row + i, col] != first)
                        {
                            win = false;
                            break;
                        }
                    }
                    if (win)
                    {
                        return first;
                    }
                }
            }
        }

        // Check diagonals (top-left to bottom-right)
        for (int row = 0; row <= boardSize - winLength; row++)
        {
            for (int col = 0; col <= boardSize - winLength; col++)
            {
                PlayerType first = board.Board[row, col];
                if (first != PlayerType.None)
                {
                    bool win = true;
                    for (int i = 1; i < winLength; i++)
                    {
                        if (board.Board[row + i, col + i] != first)
                        {
                            win = false;
                            break;
                        }
                    }
                    if (win)
                    {
                        return first;
                    }
                }
            }
        }

        // Check diagonals (top-right to bottom-left)
        for (int row = 0; row <= boardSize - winLength; row++)
        {
            for (int col = winLength - 1; col < boardSize; col++)
            {
                PlayerType first = board.Board[row, col];
                if (first != PlayerType.None)
                {
                    bool win = true;
                    for (int i = 1; i < winLength; i++)
                    {
                        if (board.Board[row + i, col - i] != first)
                        {
                            win = false;
                            break;
                        }
                    }
                    if (win)
                    {
                        return first;
                    }
                }
            }
        }

        // Check for draw (no empty cells)
        for (int row = 0; row < boardSize; row++)
        {
            for (int col = 0; col < boardSize; col++)
            {
                if (board.Board[row, col] == PlayerType.None)
                {
                    return null; // Game still in progress
                }
            }
        }

        return PlayerType.Draw;
    }

    public GameBoardState MakeMove(GameBoardState state, int row, int col)
    {
        // If game is not in playing state or cell is already occupied
        if (state.GameStatus != GameStatus.Playing || state.Board[row, col] != PlayerType.None)
        {
            return state;
        }

        // Create new state with the move
        var newState = new GameBoardState
        {
            Board = (PlayerType[,])state.Board.Clone(),
            CurrentPlayer = state.CurrentPlayer,
            GameStatus = state.GameStatus,
            Winner = state.Winner,
            Players = state.Players,
            MoveHistory = new List<Move>(state.MoveHistory),
            UndoStack = state.UndoStack,
            RedoStack = state.RedoStack,
            SessionStats = state.SessionStats
        };

        // Apply the move
        newState.Board[row, col] = state.CurrentPlayer;

        // Add move to history
        newState.MoveHistory.Add(new Move
        {
            Player = state.CurrentPlayer,
            Position = new[] { row, col },
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
        });

        // Check for winner or draw
        var winner = CheckWinner(newState);
        if (winner != null)
        {
            newState.GameStatus = winner == PlayerType.Draw ? GameStatus.Draw : GameStatus.Won;
            newState.Winner = winner == PlayerType.Draw ? null : winner;
        }

        // Switch player if game is still in progress
        if (newState.GameStatus == GameStatus.Playing)
        {
            newState.CurrentPlayer = state.CurrentPlayer == PlayerType.X ? PlayerType.O : PlayerType.X;
        }

        return newState;
    }

    public GameBoardState UndoMove(GameBoardState state)
    {
        // If no moves to undo or game is won/drawn, return current state
        if (state.MoveHistory.Count == 0 || state.GameStatus != GameStatus.Playing)
        {
            return state;
        }

        // Get the last move
        var lastMove = state.MoveHistory[^1];
        var row = lastMove.Position[0];
        var col = lastMove.Position[1];

        // Create new state with the move undone
        var newState = new GameBoardState
        {
            Board = (PlayerType[,])state.Board.Clone(),
            CurrentPlayer = state.CurrentPlayer,
            GameStatus = state.GameStatus,
            Winner = state.Winner,
            Players = state.Players,
            MoveHistory = new List<Move>(state.MoveHistory),
            UndoStack = new List<Move>(state.UndoStack),
            RedoStack = new List<Move>(state.RedoStack),
            SessionStats = state.SessionStats
        };

        // Undo the move
        newState.Board[row, col] = PlayerType.None;
        newState.MoveHistory.RemoveAt(newState.MoveHistory.Count - 1);
        newState.UndoStack.Add(lastMove);

        // Switch back to the player who made the move
        newState.CurrentPlayer = lastMove.Player;

        return newState;
    }

    public GameBoardState RedoMove(GameBoardState state)
    {
        // If no moves to redo or game is won/drawn, return current state
        if (state.UndoStack.Count == 0 || state.GameStatus != GameStatus.Playing)
        {
            return state;
        }

        // Get the last undone move
        var moveToRedo = state.UndoStack[^1];
        var row = moveToRedo.Position[0];
        var col = moveToRedo.Position[1];

        // Create new state with the move redone
        var newState = new GameBoardState
        {
            Board = (PlayerType[,])state.Board.Clone(),
            CurrentPlayer = state.CurrentPlayer,
            GameStatus = state.GameStatus,
            Winner = state.Winner,
            Players = state.Players,
            MoveHistory = new List<Move>(state.MoveHistory),
            UndoStack = new List<Move>(state.UndoStack),
            RedoStack = new List<Move>(state.RedoStack),
            SessionStats = state.SessionStats
        };

        // Redo the move
        newState.Board[row, col] = moveToRedo.Player;
        newState.MoveHistory.Add(moveToRedo);
        newState.UndoStack.RemoveAt(newState.UndoStack.Count - 1);

        // Check for winner or draw
        var winner = CheckWinner(newState);
        if (winner != null)
        {
            newState.GameStatus = winner == PlayerType.Draw ? GameStatus.Draw : GameStatus.Won;
            newState.Winner = winner == PlayerType.Draw ? null : winner;
        }

        // Switch player if game is still in progress
        if (newState.GameStatus == GameStatus.Playing)
        {
            newState.CurrentPlayer = moveToRedo.Player == PlayerType.X ? PlayerType.O : PlayerType.X;
        }

        return newState;
    }
}

File: src/Po.TicTac.Web/src/services/GameLogicService.ts
import {
  GameBoardState,
  PlayerType,
  GameStatus,
  Move,
  Player,
  BOARD_SIZE,
  WIN_LENGTH,
  createEmptyBoard,
} from '../types';

/**
 * Game Logic Service - Singleton Pattern
 * Handles all game state management and move validation.
 * Implements Strategy pattern for win checking directions.
 */
class GameLogicServiceClass {
  /**
   * Creates initial game state with empty board
   * Factory Method pattern for state creation
   */
  createInitialState(
    startingPlayer: PlayerType,
    players: [Player, Player]
  ): GameBoardState {
    return {
      board: createEmptyBoard(),
      currentPlayer: startingPlayer,
      gameStatus: GameStatus.Playing,
      winner: null,
      players,
      moveHistory: [],
      winningCells: null,
    };
  }

  /**
   * Makes a move on the board
   * Returns new state (immutable pattern)
   */
  makeMove(state: GameBoardState, row: number, col: number): GameBoardState {
    // Validate move
    if (
      state.gameStatus !== GameStatus.Playing ||
      state.board[row][col] !== PlayerType.None
    ) {
      return state;
    }

    // Create new board (immutable update)
    const newBoard = state.board.map((r, ri) =>
      r.map((c, ci) => (ri === row && ci === col ? state.currentPlayer : c))
    );

    // Create new move record
    const newMove: Move = {
      player: state.currentPlayer,
      position: [row, col],
      timestamp: Date.now(),
    };

    // Check for winner
    const winResult = this.checkWinner(newBoard);
    const nextPlayer =
      state.currentPlayer === PlayerType.X ? PlayerType.O : PlayerType.X;

    return {
      ...state,
      board: newBoard,
      currentPlayer:
        winResult.winner !== null ? state.currentPlayer : nextPlayer,
      gameStatus: winResult.status,
      winner: winResult.winner,
      moveHistory: [...state.moveHistory, newMove],
      winningCells: winResult.winningCells,
    };
  }

  /**
   * Checks for a winner on the board
   * Uses directional scanning strategy
   */
  checkWinner(board: PlayerType[][]): {
    winner: PlayerType | null;
    status: GameStatus;
    winningCells: [number, number][] | null;
  } {
    // Direction vectors: [rowDelta, colDelta]
    const directions: [number, number][] = [
      [0, 1], // Horizontal
      [1, 0], // Vertical
      [1, 1], // Diagonal down-right
      [-1, 1], // Diagonal up-right
    ];

    for (let row = 0; row < BOARD_SIZE; row++) {
      for (let col = 0; col < BOARD_SIZE; col++) {
        const cell = board[row][col];
        if (cell === PlayerType.None) continue;

        for (const [dr, dc] of directions) {
          const winCells = this.checkLine(board, row, col, dr, dc, cell);
          if (winCells) {
            return {
              winner: cell,
              status: GameStatus.Won,
              winningCells: winCells,
            };
          }
        }
      }
    }

    // Check for draw (no empty cells)
    const hasEmpty = board.some((row) =>
      row.some((cell) => cell === PlayerType.None)
    );
    if (!hasEmpty) {
      return { winner: null, status: GameStatus.Draw, winningCells: null };
    }

    return { winner: null, status: GameStatus.Playing, winningCells: null };
  }

  /**
   * Checks a line in a specific direction for winning sequence
   */
  private checkLine(
    board: PlayerType[][],
    startRow: number,
    startCol: number,
    rowDelta: number,
    colDelta: number,
    player: PlayerType
  ): [number, number][] | null {
    const cells: [number, number][] = [[startRow, startCol]];

    for (let i = 1; i < WIN_LENGTH; i++) {
      const newRow = startRow + i * rowDelta;
      const newCol = startCol + i * colDelta;

      // Check bounds
      if (
        newRow < 0 ||
        newRow >= BOARD_SIZE ||
        newCol < 0 ||
        newCol >= BOARD_SIZE
      ) {
        return null;
      }

      if (board[newRow][newCol] !== player) {
        return null;
      }

      cells.push([newRow, newCol]);
    }

    return cells;
  }

  /**
   * Gets available moves on the board
   */
  getAvailableMoves(board: PlayerType[][]): [number, number][] {
    const moves: [number, number][] = [];
    for (let row = 0; row < BOARD_SIZE; row++) {
      for (let col = 0; col < BOARD_SIZE; col++) {
        if (board[row][col] === PlayerType.None) {
          moves.push([row, col]);
        }
      }
    }
    return moves;
  }
}

// Singleton export
export const GameLogicService = new GameLogicServiceClass();
